server/apm/apm-server/apm-engine/src/main/java/io/holoinsight/server/apm/engine/elasticsearch/storage/impl/SpanEsStorage.java

```diff
@@ -105,6 +106,55 @@ public TraceBrief queryBasicTraces(final String tenant, String serviceName,
   public Trace queryTrace(String tenant, long start, long end, String traceId, List<Tag> tags)
       throws IOException {
     Trace trace = new Trace();
+    trace.setSpans(querySpan(tenant, start, end, traceId, tags));
+
+    List<Span> sortedSpans = new LinkedList<>();
+    if (CollectionUtils.isNotEmpty(trace.getSpans())) {
+      List<Span> rootSpans = findRoot(trace.getSpans());
+      if (CollectionUtils.isNotEmpty(rootSpans)) {
+        rootSpans.forEach(span -> {
+          List<Span> childrenSpan = new ArrayList<>();
+          childrenSpan.add(span);
+          findChildren(trace.getSpans(), span, childrenSpan);
+          sortedSpans.addAll(childrenSpan);
+        });
+      }
+    }
+    trace.getSpans().clear();
+    sortedSpans.forEach(span -> {
+      if (StringUtils.isEmpty(span.getParentSpanId())
+          && CollectionUtils.isNotEmpty(span.getRefs())) {
+        span.setParentSpanId(span.getRefs().get(0).getParentSpanId());
+      }
+    });
+    trace.getSpans().addAll(sortedSpans);
+    return trace;
+  }
+
+  @Override
+  public List<TraceTree> queryTraceTree(String tenant, long start, long end, String traceId,
```

Comment:
There are some new confusing function names, add some comments to help readers understand them: `queryTraceTree`, `findRoot1`, `findChildren1`...